File: frontend/src/features/twin/types/connectedAgent.ts
export type ConnectedAgent = {
  device_id: string;
  hostname?: string | null;
  os?: string | null;
  os_version?: string | null;
  arch?: string | null;
  agent_version?: string | null;
  status?: string | null;
  public_ip?: string | null;
  network_type?: string | null;
  listening_count?: number | null;
  established_count?: number | null;
  presence?: string | null;
  idle_sec?: number | null;
  app_switches?: number | null;
  last_seen_at?: string | null;
  connected: boolean;
  client_details?: Record<string, unknown>;
  network_summary?: Record<string, unknown>;
  action_summary?: Record<string, unknown>;
};

export type ConnectedAgentListResponse = {
  items: ConnectedAgent[];
  total: number;
  connected_within_sec: number;
};

export type AgentTelemetryEvent = {
  event_id: string;
  device_id: string;
  type: string;
  ts?: string | null;
  payload: Record<string, unknown>;
};

export type AgentTelemetryEventListResponse = {
  items: AgentTelemetryEvent[];
  total: number;
  device_id: string;
};

File: backend/app/features/twin/schemas/connected_agent.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class ConnectedAgentRead(BaseModel):
    device_id: str
    hostname: Optional[str] = None
    os: Optional[str] = None
    os_version: Optional[str] = None
    arch: Optional[str] = None
    agent_version: Optional[str] = None
    status: Optional[str] = None
    public_ip: Optional[str] = None
    network_type: Optional[str] = None
    listening_count: Optional[int] = None
    established_count: Optional[int] = None
    presence: Optional[str] = None
    idle_sec: Optional[int] = None
    app_switches: Optional[int] = None
    last_seen_at: Optional[datetime] = None
    connected: bool = False
    client_details: dict[str, Any] = Field(default_factory=dict)
    network_summary: dict[str, Any] = Field(default_factory=dict)
    action_summary: dict[str, Any] = Field(default_factory=dict)


class ConnectedAgentListResponse(BaseModel):
    items: list[ConnectedAgentRead] = Field(default_factory=list)
    total: int = 0
    connected_within_sec: int = 300


class AgentEventRead(BaseModel):
    event_id: str = ""
    device_id: str = ""
    type: str = ""
    ts: Optional[str] = None
    payload: dict[str, Any] = Field(default_factory=dict)


class AgentEventListResponse(BaseModel):
    items: list[AgentEventRead] = Field(default_factory=list)
    total: int = 0
    device_id: str

